alpa/device_mesh.py

```diff
@@ -1780,12 +1783,32 @@ def establish_nccl_group(self, src_mesh_id: int, dst_mesh_id: int):
         dst_mesh = self.meshes[dst_mesh_id]
         device_strs = OrderedSet(src_mesh.device_strs + dst_mesh.device_strs)
         cg = CollectiveGroup(device_strs, src_mesh, dst_mesh)
+        if instantiate:
+            if global_config.eagerly_create_communicators:
+                cg.instantiate_now()
+            else:
+                cg.instantiate()
+        self.collective_groups[src_mesh_id][dst_mesh_id] = cg
+        self.collective_groups[dst_mesh_id][src_mesh_id] = cg
+
+    def instantiate_nccl_group(self,
+                               src_mesh_id: int,
+                               dst_mesh_id: int):
+        cg = self.collective_groups[src_mesh_id][dst_mesh_id]
         if global_config.eagerly_create_communicators:
             cg.instantiate_now()
         else:
             cg.instantiate()
-        self.collective_groups[src_mesh_id][dst_mesh_id] = cg
-        self.collective_groups[dst_mesh_id][src_mesh_id] = cg
+
+    def instantiate_all(self):
+        for src_id in range(len(self)):
+            for dst_id in range(src_id + 1, len(self)):
+                cg = self.collective_groups[src_id][dst_id]
+                if cg and not cg.instantiated:
+                    if global_config.eagerly_create_communicators:
+                        cg.instantiate_now()
+                    else:
+                        cg.instantiate()
```

Comment:
This pattern repeats three times. Can we avoid it?

Reply:
Now all of them call the same internal method